tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -2451,53 +2461,123 @@ Status ConvertExpandDims(OpConverterParams* params) {
   // ExpandDim's ability to add an axis at end of the shape.
   int trt_axis;
   TF_RETURN_IF_ERROR(ConvertAxis(axis[0], dims.nbDims + 1, node_def.name(),
-                                 /*use_implicit_batch=*/true, &trt_axis));
+                                 params->use_implicit_batch, &trt_axis));
   if (params->validation_only) return Status::OK();
-
-  // ExpandDims: Insert new dim of size 1.
-  input_dims.insert(input_dims.begin() + trt_axis, 1);
-  // Reshape tensor.
-  nvinfer1::Dims new_dims;
-  TF_RETURN_IF_ERROR(TensorShapeArrayToTrtDims(input_dims, &new_dims));
   nvinfer1::ITensor* output_tensor = nullptr;
-  TF_RETURN_IF_ERROR(params->converter->PrepareTensorForShape(
-      input_tensor, new_dims, /*validation_only=*/false, &output_tensor));
+
+  if (!params->use_implicit_batch && !HasStaticShape(input_dims)) {
+    TF_RETURN_IF_ERROR(params->converter->DynamicExpandDims(
+        input_tensor.tensor(), dims, trt_axis, params, &output_tensor));
+  } else {
+    // ExpandDims: Insert new dim of size 1.
+    input_dims.insert(input_dims.begin() + trt_axis, 1);
+    // Reshape tensor.
+    nvinfer1::Dims new_dims;
+    TF_RETURN_IF_ERROR(TensorShapeArrayToTrtDims(input_dims, &new_dims));
+    TF_RETURN_IF_ERROR(params->converter->PrepareTensorForShape(
+        input_tensor, new_dims, /*validation_only=*/false, &output_tensor));
+  }
   params->outputs->push_back(TRT_TensorOrWeights(output_tensor));
   return Status::OK();
 }
 
+// Gets the shape of input, applies slice ops on the shape, and optionally
+// inserts extra dimensions between the slices. Returns the modified shape
+// tensor.
+//
+// Before each slice we can insert an extra dim if the corresponding extra_dims
+// element is not negative. Extra dims can have slices.size()+1 elements, in
+// order to insert a dimension after the last slice.
+//
+// Parameters:
+// input - input tensor
+// slices - [start, end) pairs of slices
+// params - conversion parameters
+// output - shape of input tensor sliced and expanded
+// extra_dims - if extra_dims[i] >= 0 then insert that dim before slice[i]
+Status Converter::SliceShape(nvinfer1::ITensor* input,
+                             std::vector<std::pair<int, int>> slices,
+                             OpConverterParams* params,
+                             nvinfer1::ITensor** output,
+                             std::vector<int> extra_dims) {
+  *output = nullptr;
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+  if (params->validation_only) {
+    return Status::OK();
```

Comment:
As clarified offline: you suggest to return an internal error because one should not call this function in the validation phase. This can be easily done, since this function is currently not called in validation phase.